aten/src/ATen/native/SharedReduceOps.h

```diff
@@ -204,20 +204,20 @@ struct AbsMaxOps {
 #endif
 };
 
-template <typename acc_t>
+template <typename acc_t, typename data_t, typename out_t>
 struct NormOps {
   acc_t norm_;
 
-  inline C10_DEVICE acc_t reduce(acc_t acc, acc_t data, int64_t /*idx*/) const {
+  inline C10_DEVICE acc_t reduce(acc_t acc, data_t data, int64_t /*idx*/) const {
     return acc + compat_pow(std::abs(data), norm_);
```

Comment:
Note that here you'd actually be exponentiating los precision data value (unless compat_pow is actually overloaded for BFloat16 to use fp32, but there is no way I would know it just reading the code here), and accuracy loss can indeed happen.

Reply:
This PR is just trying to "Clearly define accumulate type, data type and output type for shared reduce kernels" without involving much `BFloat16` code. And this PR is a general feature. Actually, we are planning to prepare more PRs to enable BFloat16 mathematical operations in  [BFloat16-inl.h](https://github.com/pytorch/pytorch/blob/master/c10/util/BFloat16-inl.h#L264) (same as `exp`, `log`).